working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1748,65 +1746,196 @@ types of all of the case expressions.
+        matching context, the variable is final if the variable pattern is
+        marked `final` and is not otherwise.
+
+#### Scope
 
-*   **Variable**: When not in an assignment context, introduces a variable whose
-    name is the pattern's identifier. In a declaration context, the variable is
-    final if the surrounding `patternVariableDeclaration` has a `final`
-    modifier. In a matching context, the variable is final if the variable
-    pattern is marked `final` and is not otherwise.
+The variables defined by a pattern and its subpatterns (its pattern variable
+set, defined above), are introduced into a scope based on where the pattern
+appears:
 
-[2473]: https://github.com/dart-lang/language/issues/2473
+*   **Pattern variable declaration**: The scope enclosing the variable
+    declaration statement. *This will be either a function body scope or a block
+    scope.*
+
+    The *initializing expression* for every variable in the pattern is the
+    pattern variable declaration's initializer. *This means all variables
+    defined by a pattern are in scope beginning at the top of the surrounding
+    block or function body, but it is a compile-time error to refer to them
+    until after the pattern variable declaration's initializer:*
+
+    ```dart
+    const c = 1;
+
+    f() {
+      print(c);
+      //    ^ Error: Refers to C declared below:
+
+      var [c] = c;
+      //        ^ Error: Not initialized yet.
+
+      print(c);
+      //    ^ OK.
+    }
+    ```
+
+*   **Pattern-for statement**: Scoping follows the normal for and for-in
+    statement scoping rules where the variable (now variables) are bound in a
+    new scope for each loop iteration. All pattern variables are in the same
+    scope. They are considered initialized after the for loop initializer
+    expression.
+
+*   **Pattern assignment**: An assignment only assigns to existing variables
+    and does not bind any new ones.
+
+*   **Switch statement**, **switch expression**, **if-case statement**: Each
+    `caseHead` introduces a new *case scope* which is where the variables
+    defined by that case's pattern are bound.
+
+    There is no *initializing expression* for the variables in a case pattern,
+    but they are considered initialized after the entire case pattern, before
+    the guard expression if there is one. *However, all pattern variables are
+    in scope in the entire pattern:*
+
+    ```dart
+    const c = 1;
+    switch (1) {
+      case [var c, == c]
+        //            ^ Error: In scope but not initialized.
+        //              (Also an error because `c` is not a constant.)
+            when c == 2:
+        //       ^ OK.
+        print(c);
+        //    ^ OK.
+    }
+    ```
 
-The scope where a pattern's variables are declared and available for use depends
-on the construct that contains the pattern:
+    The guard expression is evaluated in its case's case scope.
```

Comment:
This never actually says that the `then` branch of an `if`/`case` is evaluated in the case scope of the `case`.
(The switch case is handled in the next section, the `if` isn't.)

Reply:
Heh, oops. Added.